cmd/collector/app/zipkin/zipkin_json.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zipkin
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/uber/jaeger/thrift-gen/zipkincore"
+)
+
+type endpoint struct {
+	ServiceName string `json:"serviceName"`
+	IPv4        string `json:"ipv4"`
+	IPv6        string `json:"ipv6"`
+	Port        int16  `json:"port"`
+}
+type annotation struct {
+	Endpoint  endpoint `json:"endpoint"`
+	Value     string   `json:"value"`
+	Timestamp int64    `json:"timestamp"`
+}
+type binaryAnnotation struct {
+	Endpoint endpoint `json:"endpoint"`
+	Key      string   `json:"key"`
+	Value    string   `json:"value"`
```

Comment:
why did [Adrian refer](https://github.com/openzipkin/zipkin/issues/1499#issuecomment-322213158) to `oneOf`? Is this the right JSON format?